fem/coefficient.cpp

```diff
@@ -52,6 +125,159 @@ double GridFunctionCoefficient::Eval (ElementTransformation &T,
    return GridF -> GetValue (T, ip, Component);
 }
 
+void GridFunctionCoefficient::Eval(const FiniteElementSpace &fes,
+                                   const IntegrationRule &ir,
+                                   Vector &qcoeff)
+{
+   const FiniteElementSpace &coeff_fes = *GridF->FESpace();
+   const int ne = fes.GetMesh()->GetNE();
+   const int nq = ir.GetNPoints();
+   const int dim = fes.GetMesh()->Dimension();
+   qcoeff.SetSize(nq * ne);
+   if (UsesTensorBasis(coeff_fes))
+   {
+      const DofToQuad &maps = coeff_fes.GetFE(0)->GetDofToQuad(ir, DofToQuad::TENSOR);
+      const int D1D = maps.ndof;
+      const int Q1D = maps.nqpt;
+      const Operator *r = coeff_fes.GetElementRestriction(ElementDofOrdering::LEXICOGRAPHIC);
+      Vector GridE(nq*ne);
+      r->Mult(*GridF, GridE);
+      if (dim == 1)
+      {
+         auto b = Reshape(maps.B.Read(), Q1D, D1D);
+         auto x = Reshape(GridE.Read(), D1D, ne);
+         auto C = Reshape(qcoeff.Write(), Q1D, ne);
+
+         MFEM_FORALL_3D(e, ne, Q1D, 1, 1,
+         {
+            MFEM_SHARED double B[16][16];
+            for (int d = 0; d < D1D; d++)
+            {  
+               MFEM_FOREACH_THREAD(q,x,Q1D)
+               {
+                  B[q][d] = b(q,d);
+               }
+            }
+            MFEM_SYNC_THREAD;
+            MFEM_FOREACH_THREAD(qx,x,Q1D)
+            {
+               double qval = 0.0;
+               for (int dx = 0; dx < D1D; ++dx)
+               {
+                  qval += B[qx][dx] * x(dx,e);
+               }
+               C(qx,e) = qval;
+            }
+         });
+      }
+      else if (dim == 2)
+      {
+         auto b = Reshape(maps.B.Read(), Q1D, D1D);
+         auto x = Reshape(GridE.Read(), D1D, D1D, ne);
+         auto C = Reshape(qcoeff.Write(), Q1D, Q1D, ne);
+
+         MFEM_FORALL_3D(e, ne, Q1D, Q1D, 1,
+         {
+            MFEM_SHARED double B[16][16];
+            MFEM_FOREACH_THREAD(d,y,D1D)
+            {
+               MFEM_FOREACH_THREAD(q,x,Q1D)
+               {
+                  B[q][d] = b(q,d);
+               }
+            }
+            MFEM_SYNC_THREAD;
+            MFEM_FOREACH_THREAD(qx,x,Q1D)
+            {
+               MFEM_FOREACH_THREAD(qy,y,Q1D)
+               {
+                  double qval = 0.0;
+                  for (int dy = 0; dy < D1D; ++dy)
+                  {
+                     const double By = B[qy][dy];
+                     for (int dx = 0; dx < D1D; ++dx)
+                     {
+                        qval += B[qx][dx]*By * x(dx,dy,e);
+                     }
+                  }
+                  C(qx,qy,e) = qval;
+               }
+            }
+         });
+      }
+      else if (dim == 3)
+      {
+         auto b = Reshape(maps.B.Read(), Q1D, D1D);
+         auto x = Reshape(GridE.Read(), D1D, D1D, D1D, ne);
+         auto C = Reshape(qcoeff.Write(), Q1D, Q1D, Q1D, ne);
+
+         MFEM_FORALL_3D(e, ne, Q1D, Q1D, Q1D,
```

Comment:
Very excited for this! However; I thought MFEM had functions to sample grid-functions at quadrature points? The ``QuadratureInterpolator`` . Could that be applied here?

Reply:
My understanding is that the `QuadInterpolator` is for the mesh `GridFunction`.

Reply:
Maybe not...

Reply:
Good suggestion, thank you.